### PR TITLE
[PriceHistory] Combine migrations into one

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| License         | MIT                                                          |

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20230310160512.php [hunk 1]
```diff
@@ -20,21 +20,70 @@ final class Version20230310160512 extends AbstractMigration
...
     public function up(Schema $schema): void
     {
+        $this->skipIf(
+            $schema->hasTable('sylius_channel_pricing_log_entry')
+            , 'Skipping migration: Sylius price history migrations were previously executed.'
+        );
+
         $this->addSql('CREATE TABLE sylius_channel_pricing_log_entry (id INT AUTO_INCREMENT NOT NULL, channel_pricing_id INT NOT NULL, price INT NOT NULL, original_price INT DEFAULT NULL, logged_at DATETIME NOT NULL, INDEX IDX_77181A53EADFFE5 (channel_pricing_id), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
         $this->addSql('ALTER TABLE sylius_channel_pricing_log_entry ADD CONSTRAINT FK_77181A53EADFFE5 FOREIGN KEY (channel_pricing_id) REFERENCES sylius_channel_pricing (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE sylius_channel_pricing ADD lowest_price_before_discount INT DEFAULT NULL');
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20230310160512.php]: These `ALTER(S)` below should also be executed conditionally, as someone may already have these columns

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20230310160512.php [hunk 2]
```diff
@@ -20,21 +20,70 @@ final class Version20230310160512 extends AbstractMigration
...
+        $this->addSql('ALTER TABLE sylius_channel ADD lowest_price_for_discounted_products_checking_period INT DEFAULT 30 NOT NULL');
+        $this->addSql('CREATE TABLE sylius_channel_excluded_taxons (channel_id INT NOT NULL, taxon_id INT NOT NULL, INDEX IDX_3574E1E972F5A1AA (channel_id), INDEX IDX_3574E1E9DE13F470 (taxon_id), PRIMARY KEY(channel_id, taxon_id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20230310160512.php]: I would be in favor of not creating this table, as it is removed a few lines below
https://github.com/Sylius/Sylius/pull/14903/files#diff-129564f0d3ce19ac93be8b110830e662c98c41fa7348c6b76873a203e6117ec8R47

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20230310160512.php [hunk 3]
```diff
@@ -20,21 +20,70 @@ final class Version20230310160512 extends AbstractMigration
...
     {
+        $this->skipIf(
+            $schema->hasTable('sylius_channel_pricing_log_entry')
+            , 'Skipping migration: Sylius price history migrations were previously executed.'
+        );
+
         $this->addSql('CREATE TABLE sylius_channel_pricing_log_entry (id INT AUTO_INCREMENT NOT NULL, channel_pricing_id INT NOT NULL, price INT NOT NULL, original_price INT DEFAULT NULL, logged_at DATETIME NOT NULL, INDEX IDX_77181A53EADFFE5 (channel_pricing_id), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
         $this->addSql('ALTER TABLE sylius_channel_pricing_log_entry ADD CONSTRAINT FK_77181A53EADFFE5 FOREIGN KEY (channel_pricing_id) REFERENCES sylius_channel_pricing (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE sylius_channel_pricing ADD lowest_price_before_discount INT DEFAULT NULL');
+        $this->addSql('ALTER TABLE sylius_channel ADD lowest_price_for_discounted_products_checking_period INT DEFAULT 30 NOT NULL');
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20230310160512.php]: Same as:
https://github.com/Sylius/Sylius/pull/14903/files#r1153325226

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20230331091850.php [hunk 4]
```diff
@@ -16,25 +16,33 @@
...
+        $this->addSql('ALTER TABLE sylius_channel ADD channel_price_history_config_id INT DEFAULT NULL');
         $this->addSql('ALTER TABLE sylius_channel ADD CONSTRAINT FK_16C8119E75F20EAE FOREIGN KEY (channel_price_history_config_id) REFERENCES sylius_channel_price_history_config (id) ON DELETE CASCADE');
         $this->addSql('CREATE UNIQUE INDEX UNIQ_16C8119E75F20EAE ON sylius_channel (channel_price_history_config_id)');
+        $this->addSql('ALTER TABLE sylius_channel_pricing ADD lowest_price_before_discount INT DEFAULT NULL');
+
+        /** Create an initial log state based on the price of products at the time of migration processing */
+        $this->addSql('INSERT INTO `sylius_channel_pricing_log_entry` (`channel_pricing_id`, `price`, `original_price`, `logged_at`) SELECT `id`, `price`, `original_price`, NOW() FROM `sylius_channel_pricing`');
     }
 
     public function postUp(Schema $schema): void
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20230331091850.php:48]: Either move the content of this method to `::up` or change `SELECT id from sylius_channel` to `SELECT id from sylius_channel WHERE channel_price_history_config_id IS NULL`, so it won't try to recreate existing configs

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20230331091850.php [hunk 5]
```diff
@@ -16,30 +16,38 @@
 use Doctrine\DBAL\Schema\Schema;
 use Doctrine\Migrations\AbstractMigration;
 
-final class Version20230329093932 extends AbstractMigration
+final class Version20230331091850 extends AbstractMigration
 {
     public function getDescription(): string
     {
-        return 'Add ChannelPriceHistoryConfig';
+        return 'Create ChannelPricingLogEntry, ChannelPriceHistoryConfig, table of excluded taxons from showing the lowest price and add channel_price_history_config_id column to sylius_channel table';
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20230331091850.php]: IMO it should be enough:
```suggestion
        return 'Initialize Price History';
```

**Review by GSadee — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @TheMilek!

**Pull request merged**